### ISSUE TITLE
Incorrect calculation of the formation energy

### ISSUE THREAD

**Issue by raimis** [openmm/spice-dataset#28]
@peastman I have checked the latest SPICE dataset you shared with us. The is a problem with the calculation of the formation energies.

The formation energy of molecule is a negative energy change of moving the constituting atoms to infinite distance form each other. Also, this assumes that the electrons redistribute to their ground state in the process. For neutral molecules, it is trivial. One just need to subtract the sum of energies of isolated neutral atoms. For charged molecules, one need to ensure that electrons are neither create nor annihilated, and their at the ground state.

The current algorithm assumes that the electrons in the isolated atoms are distributed according to their formal charges. For example, the energy of the isolated atoms for `[N-:1]=[N+:3]=[N-:2]` molecule (ID `135041958` in the dataset) is computed as:
```python
atom_energy[('N', -1)] + atom_energy[('N', -1)] + atom_energy[('N', 1)]
-163.29052361673166
```
This assumes that two `N` atoms are missing one electron and one has an extra one. However, the ground state (i.e. the lowest-energy state) is when two `N` atoms are neutral and one has an extra electron:
```python
atom_energy[('N', 0)] + atom_energy[('N', 0)] + atom_energy[('N', -1)]
-163.84884136280493
```
So, the formation energies of all the zwitterionic molecules and some charged ones are computed incorrectly.

**Comment by peastman** [comment]
@YuezhiMao any thoughts on this?

For our purposes the difference doesn't really matter.  The important thing is just that we subtract off the large constant energy that makes training difficult.  Possibly we just need to call it something different from "formation energy"?

**Comment by YuezhiMao** [comment]
I think what @raimis said regarding formation energy is correct. For charged species, we will need to assign the charge to one of the atoms such that **the sum of atomic energies is the lowest**. It might be important to follow this for all charged species since otherwise we might be artificially creating some bias due to inconsistent choice of the atomic states. I consulted someone who is experienced with DFT benchmark studies in the Head-Gordon group today, and they basically suggested the same. Since this just involves combining the atomic energies for the charged species in a different way, I guess it won't be too difficult to implement?

**Comment by peastman** [comment]
Can you enumerate which charge states we need to consider for the individual atoms?  Here's the list of elements found in the dataset: Br, C, Ca, Cl, F, H, I, K, Li, Mg, N, Na, O, P, S.  For each one, can you list the charges it might have?

**Comment by YuezhiMao** [comment]
May I get an idea on what kind of charged species are in the dataset? Are there any open-shell species (radicals)?

**Comment by raimis** [comment]
Everything should be in a closed shell and the total charges are just `-1`, `0`, and `+1`.

**Comment by peastman** [comment]
The total charge can be less than -1 or more than +1.  For example, a dipeptide consisting of two charged residues, or a pair of ions with the same charge.

If we can enumerate the charge states for each element, I think the following algorithm will give the result we want?

- Start by assuming every atom ends up with the charge for which its energy is minimum.
- While the total charge is too positive:
  - Find the atom for which decreasing its charge has the minimum energy cost and decrease it.
- While the total charge is too negative:
  - Find the atom for which increasing its charge has the minimum energy cost and increase it.

**Comment by peastman** [comment]
Using the charge states currently in the script

https://github.com/openmm/spice-dataset/blob/40795cec4226a1fa972aaae5e00b8dd61c9af5bb/downloader/downloader.py#L10-L16

the above algorithm seems to produce reasonable results.  In about 85-90% of cases it produces identical energies to using formal charges.  In the others, it reduces the energy slightly.  For example, in the dimer

```
[H]C([H])(I)C([H])([H])I.[H]C([H])([H])C(=O)[O-]
```

it moves the negative charge from an O to an I, reducing the energy from -900.7906771266157 to -900.8519328099412.  Or on this one

```
[H]N(C(=O)C([H])([H])[H])C([H])([H])[H].[H]N([H])C(N([H])C([H])([H])[H])=[N+]([H])[H]
```

it moves the positive charge from N to C, reducing the energy from -489.91357824345783 to -490.03175902857174.  Do those seem like correct things for it to do?

**Comment by peastman** [comment]
Ok, I'll add them.  Thanks!

**Comment by peastman** [comment]
The fix is in #29.

**Comment by raimis** [comment]
Yes, the algorithm is good.
